### ISSUE TITLE
MapLoader.setMap should copy the map

### ISSUE THREAD

**Issue by eeverman** [eeverman/andhow#554]
**Describe the bug**
The [MapLoader.setMap](https://github.com/eeverman/andhow/blob/69af75fc21aaf7194ecbf5df7bb62613755374c7/andhow-core/src/main/java/org/yarnandtail/andhow/load/MapLoader.java#L21) method currently keeps the passed map.  Instead, this method should copy the values in the map to its own map, similar to how the [KeyValuePairLoader.setKeyValuePairs](https://github.com/eeverman/andhow/blob/69af75fc21aaf7194ecbf5df7bb62613755374c7/andhow-core/src/main/java/org/yarnandtail/andhow/load/KeyValuePairLoader.java#L56) behaves.

If an empty or null map is passed, the existing Map should be emptied and this behavior should be noted in the javadocs.

**Comment by alex-kar** [comment]
@eeverman Can you please assign this one to me?

**Comment by alex-kar** [comment]
In KeyValuePairLoader `keyValuePairs` field initialized as empty array list
```
private List<String> keyValuePairs = new ArrayList();
```
and inside `setKeyValuePairs` just cleared if parameter list is empty or null.

In MapLoader#setMap map is not initialized (null).
```
protected Map<?, ?> map;
```
My thought was about to initialize map and clear inside `setMap`,
but `map` field has access modifier `protected` and used in different classes when logic based on `null`.
Can you clarify what the expected behavior in `setMap` ?
In case when empty or null map passed should I check current value for `map` and keep it null if it's null or clear if map initialized?

Btw, setters in KeyValuePairLoader
`setKeyValuePairs(List<String> keyValuePairs)`
`setKeyValuePairs(String... keyValuePairs)` behaves differently, second one does not clear list before add new values. 
And both of them ignore empty or null parameters.

**Comment by eeverman** [comment]
Yes, @alex-kar, all of these loaders are slightly wrong on these methods.  I had opened a separate ticket to straighten that out:  #555.  The logic of a `set` method here should be (from 555) : 

> - Completely replace values when set is called
> - Set the associated collection to empty when set is called with null or empty

This code has been this way a long time and works because these loaders are only used one time, then discarded.  The code and the javadocs do make it hard to understand what the intention is, however, so it would be good to fix this all.

Regarding initiation, I think it makes sense to use a final field and initialize the collection variable in the declaration (as `FixedValueLoader` does) and just add or clear values in it as needed.  At runtime, there are 7 or 8 loaders that are created briefly to load values, then they are garbage collected, so creating a collection that might not be used is not a big concern.

**Comment by eeverman** [comment]
Wow, @alex-kar, what a nice addition with great tests!

I did steer you wrong on one part, however.  There are two subclasses, `StdSysPropLoader` and `StdEnvVarLoader` that rely on the the map sometimes being null.  So, to revise my statement above for the MapLoader:
* `setMap(null)` should store a null
* `setMap(empty)` should result in an empty map
* `setMap([non-empty map])` should copy the values to the map

Sorry - I should have looked at the subclass before answering your question.

**Comment by eeverman** [comment]
Fixed by #564
